osdep/path-unix.c

```diff
@@ -37,36 +39,42 @@ static void path_init(void)
     char *xdg_config = getenv("XDG_CONFIG_HOME");
     char *xdg_state = getenv("XDG_STATE_HOME");
 
+    int err = 0;
     if (xdg_config && xdg_config[0]) {
-        snprintf(mpv_home, sizeof(mpv_home), "%s/mpv", xdg_config);
+        err |= MKPATH(mpv_home, "%s/mpv", xdg_config);
     } else if (home && home[0]) {
-        snprintf(mpv_home, sizeof(mpv_home), "%s/.config/mpv", home);
+        err |= MKPATH(mpv_home, "%s/.config/mpv", home);
     }
 
     // Maintain compatibility with old ~/.mpv
     if (home && home[0])
-        snprintf(old_home, sizeof(old_home), "%s/.mpv", home);
+        err |= MKPATH(old_home, "%s/.mpv", home);
 
     if (xdg_cache && xdg_cache[0]) {
-        snprintf(mpv_cache, sizeof(mpv_cache), "%s/mpv", xdg_cache);
+        err |= MKPATH(mpv_cache, "%s/mpv", xdg_cache);
     } else if (home && home[0]) {
-        snprintf(mpv_cache, sizeof(mpv_cache), "%s/.cache/mpv", home);
+        err |= MKPATH(mpv_cache, "%s/.cache/mpv", home);
     }
 
     if (xdg_state && xdg_state[0]) {
-        snprintf(mpv_state, sizeof(mpv_state), "%s/mpv", xdg_state);
+        err |= MKPATH(mpv_state, "%s/mpv", xdg_state);
     } else if (home && home[0]) {
-        snprintf(mpv_state, sizeof(mpv_state), "%s/.local/state/mpv", home);
+        err |= MKPATH(mpv_state, "%s/.local/state/mpv", home);
     }
 
     // If the old ~/.mpv exists, and the XDG config dir doesn't, use the old
     // config dir only. Also do not use any other XDG directories.
     if (mp_path_exists(old_home) && !mp_path_exists(mpv_home)) {
-        snprintf(mpv_home, sizeof(mpv_home), "%s", old_home);
-        snprintf(mpv_cache, sizeof(mpv_cache), "%s", old_home);
-        snprintf(mpv_state, sizeof(mpv_state), "%s", old_home);
+        err |= MKPATH(mpv_home, "%s", old_home);
+        err |= MKPATH(mpv_cache, "%s", old_home);
+        err |= MKPATH(mpv_state, "%s", old_home);
         old_home[0] = '\0';
     }
+
+    if (err) {
+        fprintf(stderr, "Config dir exceeds %d bytes\n", CONF_MAX);
```

Comment:
Please use "mpv: config dir exceeds..." at the message.

Reply:
That's what I initially had, but then I grepped for some other `fprintf(stderr` usage and none of them use `mpv: ` prefix - I assumed that's because mpv might be being used from another front-end.

Reply:
> mpv might be being used from another front-end.

For this we probably should never write directly to stderr without going through logging system.

Reply:
> I grepped for some other `fprintf(stderr` usage and none of them use `mpv: ` prefix

Hmm... There are not many, but indeed none seem to use "mpv:" prefix. OK, leave it.

> For this we probably should never write directly to stderr without going through logging system.

Preferably yes, but not sure if the log can be easily accessible here. We'd need to check this.

Reply:
It seems that a log is available through `mpctx` earlier in the call chain, but it gets lost in the middle at the `mp_mk_user_dir()` call (slightly cleaned up backtrace):

```
#0  path_init () at ../osdep/path-unix.c:37
#1  ?? () from /lib64/libc.so.6
#2  mp_get_platform_path_unix (talloc_ctx=0x607000000540, type=0xd04820 "home") at ../osdep/path-unix.c:82
#3  mp_get_platform_path (talloc_ctx=0x607000000540, global=0x60c000000150, type=0xd04820 "home")
    at ../options/path.c:102
#4  mp_find_user_file (talloc_ctx=0x0, global=0x60c000000150, type=0xd04820 "home", filename=0xd047e0 "")
    at ../options/path.c:131
#5  mp_mk_user_dir (global=0x60c000000150, type=0xd04820 "home", subdir=0xd047e0 "") at ../options/path.c:407
#6  mp_parse_cfgfiles (mpctx=0x61a0000000d0) at ../player/configfiles.c:66
#7  mp_initialize (mpctx=0x61a0000000d0, options=0x7fffffffdfa0) at ../player/main.c:353
#8  mpv_main (argc=2, argv=0x7fffffffdf98) at ../player/main.c:437
#9  main (argc=2, argv=0x7fffffffdf98) at ../osdep/main-fn-unix.c:5
```

But this backtrace might not be reliable at all because the path function is using `pthread_once()` - which implies that it's expecting to be called from multiple threads. Which *usually* means there are multiple places where this can be called first.

Reply:
Actually, never mind. Even if you pass the `mpctx` all the way to `mp_get_platform_path_unix` - it'll still be useless because `pthread_once()` doesn't have a context argument (the same reason why we couldn't do dynamic allocation without leaks or hacky global smuggling).

Reply:
Yeah I don't think trying to pass the logger here will work or is worth it.

Reply:
> still be useless because `pthread_once()` doesn't have a context argument (the same reason why we couldn't do dynamic allocation without leaks or hacky global smuggling).

> Yeah I don't think trying to pass the logger here will work or is worth it.

Right. Thanks for the investigation. Indeed we shouldn't try to log at this stage.